### PR TITLE
AB#2512 Config secrets via env var & config refactoring

### PR DESCRIPTION
**Pull request by datosh**
### Proposed change(s)

- refactored measurements code
  - `map[uint32][]byte` to use consistent type, makes subsequent changes easier
  - provided functions such as `Zero`, `One` and `AllBytes` to make tests more readable and faster to write
  - moved equality test from upgrade cmd to measurement type
- refactored multierr hashicorp library to use uber multierr library
- refactored config code
  - Added `New` function `WithDefaultOptions` to make usage of config in CLI much more consistent. For example, many subcommands lacked config validation.

### Checklist

- [x] Update [CHANGELOG.md](https://github.com/edgelesssys/constellation/blob/main/CHANGELOG.md)
- [x] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [x] Link to Milestone

### CODE REVIEW

**Comment by datosh** [comment]
> Sorry for drive-by review but since this touched quite a bit of code I wanted to work on, it was kinda necessary to look into here.

Not at all - thanks for taking the time to review! 🙂

File: internal/config/config.go
Comment (datosh) [reply internal/config/config.go:598]: My idea was to move the validate helper functions to a dedicated file, but leave the public API of config in its original one. I agree that it is not the cleanest cut.

One could also argue to move the private methods of config back to config.go.

What do the other reviewers think?

File: internal/attestation/measurements/measurements.go
Comment (datosh) [reply internal/attestation/measurements/measurements.go:42]: I introduced them to reduce repeating bytes and make code more readable in general. I think for the code base as it is right now it is a benefit to have these. 

I can't see how they are related to the measurements 11, 12, 13, ... you are talking about. Maybe we can have a chat once the other reviews are in..? 

> So I don't know for sure. We can also just adjust it later.

Yes, if future changes make these unnecessary we should remove them! 👍

File: internal/config/config.go
Comment (thomasten) [reply internal/config/config.go:598]: I like having the validate helper functions in another file.
Regarding Validate(), both locations are fine with me and there are reasonable arguments for both.

File: internal/config/config.go
Comment (datosh) [reply internal/config/config.go:404]: I like `config.New` as a cleaner implementation for now, and `config.NewWithOptions` as an option going forward. I'll wait for other feedback and will change the implementation if others agree as well!

File: CHANGELOG.md
Comment (thomasten) [inline CHANGELOG.md:24]: The sentence reads a bit odd. Maybe
```suggestion
Environment variable `CONSTELL_AZURE_CLIENT_SECRET_VALUE` as an alternative way to provide the configuration value `provider.azure.clientSecretValue`.
```

File: bootstrapper/internal/kubernetes/k8sapi/install_test.go
Comment (thomasten) [inline bootstrapper/internal/kubernetes/k8sapi/install_test.go:716]: good catch!

File: cli/internal/cmd/configfetchmeasurements.go
Comment (thomasten) [inline cli/internal/cmd/configfetchmeasurements.go:61]: nit: isn't the old name that doesn't shadow the package name better?

File: docs/docs/getting-started/first-steps.md
Comment (thomasten) [inline docs/docs/getting-started/first-steps.md:76]:
```suggestion
    Alternatively, you can leave `clientSecretValue` empty and provide the secret via the `CONSTELL_AZURE_CLIENT_SECRET_VALUE` environment variable.
```

File: docs/docs/getting-started/first-steps.md
Comment (thomasten) [inline docs/docs/getting-started/first-steps.md:74]: should this be moved between "Fill the values ..." and "By default, ..."?

File: internal/attestation/measurements/measurements_test.go
Comment (thomasten) [inline internal/attestation/measurements/measurements_test.go:365]: nit: not sure what our naming convention is for this kind of tests. @katexochen

File: internal/config/config.go
Comment (thomasten) [inline internal/config/config.go:595]: nit: For "seperation of concerns"-like reasons, I prefer the old variant of this function and would move the conversion to error into WithValidate(). But if you decide to keep it like this, please change the doc comment of this func.

**Review by thomasten — CHANGES_REQUESTED** [review]
Just a few smaller things. Nils already brought up good points regarding some of the high-level decisions, which I also thought about while reviewing. However, I didn't come to to a clear preference and I think that this PR is definitely an improvement to the status quo, so I'd be fine with that.

File: internal/attestation/measurements/measurements.go
Comment (malt3) [reply internal/attestation/measurements/measurements.go:42]: Yes, 11, 12 and 13 will eventually have a non-zero value

File: internal/attestation/measurements/measurements_test.go
Comment (katexochen) [reply internal/attestation/measurements/measurements_test.go:365]: In most cases we would just have called this `TestEqualTo` I guess, but for me its okay this way, too, as github.com/cweill/gotests (included in the VSCode Go ext) uses this naming convention. Maybe something we could start to adopt...

File: internal/attestation/measurements/measurements.go
Comment (katexochen) [inline internal/attestation/measurements/measurements.go:27]: nit: A common thing in Go would be to call this just `M`, like in `testing.T`, as it its the only exported type in the package and the package name is already descriptive enough...

File: internal/attestation/measurements/measurements.go
Comment (katexochen) [reply internal/attestation/measurements/measurements.go:38]: I don't like such exported utilities, especially, the name of the function isn't descriptive enough, I would `measurements.Zero` expect to return a zero value of type Measurements, not a slice of bytes. If you don't want to copy code (which isn't a bad think) or don't want to count up to 32 chars (I totally feel that), I'd suggest to provide a single utility like `AllBytes` but with a more descriptive name and comment.

File: internal/attestation/measurements/measurements.go
Comment (katexochen) [inline internal/attestation/measurements/measurements.go:60]: You could use functions instead of global vars to make this code safer. Global state is never a good idea...

File: internal/attestation/measurements/measurements_test.go
Comment (katexochen) [inline internal/attestation/measurements/measurements_test.go:332]: Why not use` bytes.Repeat` here?

File: internal/config/config.go
Comment (katexochen) [inline internal/config/config.go:33]:
```suggestion
// Measurements is a required alias since docgen is not able to work with
// types in other packages.
```

File: internal/attestation/measurements/measurements.go
Comment (datosh) [reply internal/attestation/measurements/measurements.go:38]: Renamed to `PCRWithAllBytes` and removed `Zero` and `One` functions.

File: internal/config/config.go
Comment (datosh) [reply internal/config/config.go:404]: Sorry, my answer was probably too short and therefore ambiguous: I totally agree with your suggestion! 

There is no need for the implemented flexibility when no one is using it. Implemented your suggestion 🙂

File: docs/docs/getting-started/first-steps.md
Comment (datosh) [reply docs/docs/getting-started/first-steps.md:74]: I don't have strong feelings for either - I have moved as suggested!

File: internal/attestation/measurements/measurements.go
Comment (datosh) [reply internal/attestation/measurements/measurements.go:27]: Thanks for the encouragement! I have thought about it but didn't find any instances of it in our code base, so I decided against it. I too like the short form better, so that clients do not stutter.

File: internal/attestation/measurements/measurements.go
Comment (datosh) [reply internal/attestation/measurements/measurements.go:60]: Good hint! I moved them into `DefaultsFor`. It was the only reference anyway.

File: internal/attestation/measurements/measurements_test.go
Comment (datosh) [reply internal/attestation/measurements/measurements_test.go:332]: Since this code tests `AllBytes` (now `PCRWithAllBytes`) which does the repetition I wanted to show the full expected value once.

**Comment by datosh** [comment]
Thanks for all the great feedback. Please take another look!

File: internal/attestation/measurements/measurements_test.go
Comment (katexochen) [reply internal/attestation/measurements/measurements_test.go:332]: Imo you can assume these (especially external/stdlib) dependencies to be correct, as they don't belong to your _system under test_.

File: internal/config/config.go
Comment (thomasten) [inline internal/config/config.go:334]: optional: We can try to prevent developers from accidentially using this func instead of New. I've got two suggestions:
* This func is currently only used by cdbg. If it's acceptable that cdbg validates the config, we could use New there, too, and unexport this func.
* Or we could extend the doc comment to encourage developers to use New in most cases.

File: internal/config/config.go
Comment (thomasten) [inline internal/config/config.go:357]:
```suggestion
	c, err := FromFile(fileHandler, name)
	if err != nil {
		return nil, err
	}
```

File: internal/config/config.go
Comment (thomasten) [inline internal/config/config.go:351]:
```suggestion
```

File: internal/config/config.go
Comment (thomasten) [inline internal/config/config.go:82]: Does this need to be
```suggestion
	Measurements Measurements `yaml:"measurements"`
```
because of docgen?

File: internal/config/config_doc.go
Comment (thomasten) [inline internal/config/config_doc.go:246]: don't forget to regenerate before merge

**Review by thomasten — APPROVED** [review]

File: internal/config/config_doc.go
Comment (datosh) [reply internal/config/config_doc.go:246]: Good point! Looking forward to @3u13r 's check there 😉

File: internal/config/config.go
Comment (datosh) [reply internal/config/config.go:82]: not sure why that one instance was changed.

**Pull request merged**
